File: ticketing_frontend/services/userService.ts

import { User, ManagedUser } from '../types';
import { apiFetch } from './api';

// Get all users (Admin)
export const getAllUsers = async (): Promise<ManagedUser[]> => {
  return apiFetch('/users/', { method: 'GET' });
};

// Get user by email (Admin)
export const getUserByEmail = async (email: string): Promise<ManagedUser> => {
  return apiFetch(`/users/${email}`, { method: 'GET' });
};

// Update user role (Admin)
export const updateUserRole = async (email: string, role: string): Promise<{ message: string, user: ManagedUser }> => {
  return apiFetch(`/users/${email}/role`, {
    method: 'PUT',
    body: JSON.stringify({ role }),
  });
};

// Update user associations (Admin)
export const updateUserAssociations = async (email: string, associations: string): Promise<{ message: string, user: ManagedUser }> => {
  return apiFetch(`/users/${email}/associations`, {
    method: 'PUT',
    body: JSON.stringify({ associations }),
  });
};

// Update user password by admin (Admin)
export const updateUserPasswordByAdmin = async (email: string, new_password: string): Promise<{ message: string, user: ManagedUser }> => {
  return apiFetch(`/users/${email}/password`, {
    method: 'PUT',
    body: JSON.stringify({ new_password }),
  });
};

// Delete user (Super Admin)
export const deleteUser = async (email: string): Promise<{ message: string }> => {
  return apiFetch(`/users/${email}`, { method: 'DELETE' });
};

// Update self password (Logged-in user)
export const updateSelfPassword = async (old_password: string, new_password: string): Promise<{ message: string }> => {
  return apiFetch('/users/self/password', {
    method: 'PUT',
    body: JSON.stringify({ old_password, new_password }),
  });
};

File: ticketing_backend/services/user_service.py
from models import db, User
from config import Config
from werkzeug.security import generate_password_hash


def get_user_by_id(user_id):
    return User.query.get(user_id)


def get_user_by_email(email):
    return User.query.filter_by(email=email.lower()).first()


def get_all_users():
    return User.query.all()


def update_user_role(user_email, new_role, current_admin_email):
    user = get_user_by_email(user_email)
    if not user:
        return None, "User not found."

    # Prevent anyone other than the superadmin from changing the superadmin's role
    if (
        user.email == Config.SUPER_ADMIN_EMAIL
        and current_admin_email != Config.SUPER_ADMIN_EMAIL
    ):
        return None, "Cannot alter the architect of the system!"

    if new_role == "admin" and current_admin_email != Config.SUPER_ADMIN_EMAIL:
        return None, "Only the Super Admin can assign the admin role."

    if user.email == Config.SUPER_ADMIN_EMAIL and new_role != "admin":
        return (
            None,
            f"The super admin account ({Config.SUPER_ADMIN_EMAIL}) must retain the 'admin' role.",
        )

    user.role = new_role
    db.session.commit()
    return user, None


def update_user_associations(user_email, new_associations, current_admin_email):
    user = get_user_by_email(user_email)
    if not user:
        return None, "User not found."
    if (
        user.email == Config.SUPER_ADMIN_EMAIL
        and current_admin_email != Config.SUPER_ADMIN_EMAIL
    ):
        return None, "Cannot alter the architect of the system!"

    user.associations = new_associations
    db.session.commit()
    return user, None


def update_user_password(user_email, new_password, current_admin_email):
    user = get_user_by_email(user_email)
    if not user:
        return None, "User not found."
    if (
        user.email == Config.SUPER_ADMIN_EMAIL
        and current_admin_email != Config.SUPER_ADMIN_EMAIL
    ):
        return None, "Cannot alter the architect of the system!"

    user.set_password(new_password)
    db.session.commit()
    return user, None


def update_user_password_self(user_id, old_password, new_password):
    user = User.query.get(user_id)
    if not user:
        return None, "User not found."

    if not user.check_password(old_password):
        return None, "Incorrect old password."

    user.set_password(new_password)
    db.session.commit()
    return user, None


def delete_user_by_email(user_email, current_admin_email):
    user = get_user_by_email(user_email)
    if not user:
        return None, "User not found."
    if (
        user.email == Config.SUPER_ADMIN_EMAIL
        and current_admin_email != Config.SUPER_ADMIN_EMAIL
    ):
        return None, "Cannot alter the architect of the system!"

    db.session.delete(user)
    db.session.commit()
    return True, None


def get_admins():
    return [user.email for user in User.query.filter_by(role="admin").all()]


def get_tech_admins():
    return [
        user.email
        for user in User.query.filter(
            User.role == "admin",
            User.associations.in_(
                ["bravo", "echo", "hotel", "india", "kilo", "lima", "november", "oscar"]
            ),
        ).all()
    ]


def get_maintenance_admins():
    return [
        user.email
        for user in User.query.filter(
            User.role == "admin",
            User.associations.in_(
                [
                    "delta",
                    "golf",
                    "india",
                    "juliett",
                    "lime",
                    "mike",
                    "november",
                    "oscar",
                ]
            ),
        ).all()
    ]


def get_management_admins():
    return [
        user.email
        for user in User.query.filter(
            User.role == "admin",
            User.associations.in_(
                [
                    "charlie",
                    "foxtrot",
                    "hotel",
                    "juliett",
                    "kilo",
                    "mike",
                    "november",
                    "oscar",
                ]
            ),
        ).all()
    ]


def get_user_role(user_email):
    user = get_user_by_email(user_email)
    return user.role if user else None


def get_user_associations(user_email):
    user = get_user_by_email(user_email)
    return user.associations if user else None
